File: NovilhaNutri/Core/Modelos/CicloEncerrado.swift
import Foundation

/// O que o produtor informa no dia do abate.
struct ResultadoAbate: Hashable {
    var dataAbate: Date
    var pesoFinalReal: Double
    var animaisAbatidos: Int
    /// Peso médio de carcaça em kg. Zero quando o frigorífico ainda não fechou.
    var pesoCarcacaReal: Double
    /// Concentrado realmente fornecido no ciclo, em kg de matéria natural.
    var concentradoReal: Double
    var custoReal: Double
    var precoArroba: Double
    var observacoes: String

    init(dataAbate: Date = Date(),
         pesoFinalReal: Double = 0,
         animaisAbatidos: Int = 0,
         pesoCarcacaReal: Double = 0,
         concentradoReal: Double = 0,
         custoReal: Double = 0,
         precoArroba: Double = 0,
         observacoes: String = "") {
        self.dataAbate = dataAbate
        self.pesoFinalReal = pesoFinalReal
        self.animaisAbatidos = animaisAbatidos
        self.pesoCarcacaReal = pesoCarcacaReal
        self.concentradoReal = concentradoReal
        self.custoReal = custoReal
        self.precoArroba = precoArroba
        self.observacoes = observacoes
    }
}

/// Um lote que já foi abatido, com o que foi planejado e o que de fato aconteceu.
///
/// É a memória da fazenda: cada ciclo encerrado vira base de comparação e
/// calibração para os lotes seguintes.
struct CicloEncerrado: Identifiable, Codable, Hashable {
    var id: UUID
    var loteID: UUID
    var nome: String
    var grupoGenetico: GrupoGenetico
    var sistema: SistemaCriacao

    // MARK: Plano, fotografado no momento do encerramento

    var dataInicio: Date
    var pesoInicial: Double
    var ganhoMeta: Double
    var pesoAlvo: Double
    var animaisIniciais: Int
    var pesoAcabamentoPlanejado: Double
    /// Consumo médio previsto, em kg de matéria seca por animal por dia.
    var consumoPrevistoDiario: Double
    /// Teor médio da dieta planejada (% da matéria seca).
    var ndtDietaMedia: Double
    var pbDietaMedia: Double
    /// Concentrado previsto para o ciclo inteiro, em kg de matéria natural.
    var concentradoPrevisto: Double
    var custoPrevisto: Double
    var volumosoNome: String
    var energeticoNome: String
    var proteicoNome: String

    // MARK: Realizado

    var dataAbate: Date
    var pesoFinalReal: Double
    var animaisAbatidos: Int
    /// Peso médio de carcaça, em kg. Zero quando não informado.
    var pesoCarcacaReal: Double
    /// Concentrado realmente fornecido no ciclo, em kg de matéria natural.
    var concentradoReal: Double
    var custoReal: Double
    /// Preço recebido por arroba. Zero quando não informado.
    var precoArroba: Double
    var observacoes: String

    init(id: UUID = UUID(),
         loteID: UUID,
         nome: String,
         grupoGenetico: GrupoGenetico,
         sistema: SistemaCriacao,
         dataInicio: Date,
         pesoInicial: Double,
         ganhoMeta: Double,
         pesoAlvo: Double,
         animaisIniciais: Int,
         pesoAcabamentoPlanejado: Double,
         consumoPrevistoDiario: Double,
         ndtDietaMedia: Double,
         pbDietaMedia: Double,
         concentradoPrevisto: Double,
         custoPrevisto: Double,
         volumosoNome: String,
         energeticoNome: String,
         proteicoNome: String,
         dataAbate: Date,
         pesoFinalReal: Double,
         animaisAbatidos: Int,
         pesoCarcacaReal: Double = 0,
         concentradoReal: Double = 0,
         custoReal: Double = 0,
         precoArroba: Double = 0,
         observacoes: String = "") {
        self.id = id
        self.loteID = loteID
        self.nome = nome
        self.grupoGenetico = grupoGenetico
        self.sistema = sistema
        self.dataInicio = dataInicio
        self.pesoInicial = pesoInicial
        self.ganhoMeta = ganhoMeta
        self.pesoAlvo = pesoAlvo
        self.animaisIniciais = animaisIniciais
        self.pesoAcabamentoPlanejado = pesoAcabamentoPlanejado
        self.consumoPrevistoDiario = consumoPrevistoDiario
        self.ndtDietaMedia = ndtDietaMedia
        self.pbDietaMedia = pbDietaMedia
        self.concentradoPrevisto = concentradoPrevisto
        self.custoPrevisto = custoPrevisto
        self.volumosoNome = volumosoNome
        self.energeticoNome = energeticoNome
        self.proteicoNome = proteicoNome
        self.dataAbate = dataAbate
        self.pesoFinalReal = pesoFinalReal
        self.animaisAbatidos = animaisAbatidos
        self.pesoCarcacaReal = pesoCarcacaReal
        self.concentradoReal = concentradoReal
        self.custoReal = custoReal
        self.precoArroba = precoArroba
        self.observacoes = observacoes
    }

    // MARK: - Desempenho observado

    /// Duração real do ciclo, em dias.
    var diasReais: Double {
        max(dataAbate.timeIntervalSince(dataInicio) / 86_400, 0)
    }

    var diasPlanejados: Double {
        ganhoMeta > 0 ? max(0, pesoAlvo - pesoInicial) / ganhoMeta : 0
    }

    var ganhoTotalPorAnimal: Double { max(0, pesoFinalReal - pesoInicial) }

    var ganhoTotalLote: Double { ganhoTotalPorAnimal * Double(animaisAbatidos) }

    /// Ganho médio diário observado (kg/dia).
    var ganhoRealDiario: Double {
        diasReais >= 1 ? ganhoTotalPorAnimal / diasReais : 0
    }

    /// Quanto do ganho planejado foi entregue. 1,0 = exatamente a meta.
    var aderenciaGanho: Double {
        ganhoMeta > 0 ? ganhoRealDiario / ganhoMeta : 0
    }

    /// Peso médio do animal ao longo do ciclo, usado nas contas de calibração.
    var pesoMedioCiclo: Double { (pesoInicial + pesoFinalReal) / 2 }

    /// Rendimento de carcaça observado (fração). Nulo quando não informado.
    var rendimentoReal: Double? {
        guard pesoCarcacaReal > 0, pesoFinalReal > 0 else { return nil }
        return pesoCarcacaReal / pesoFinalReal
    }

    var arrobasPorAnimal: Double {
        pesoCarcacaReal > 0 ? pesoCarcacaReal / 15 : 0
    }

    var arrobasLote: Double { arrobasPorAnimal * Double(animaisAbatidos) }

    /// Arrobas produzidas no ciclo, descontando a carcaça de entrada.
    var arrobasProduzidasLote: Double {
        guard let rendimento = rendimentoReal else { return 0 }
        let carcacaInicial = pesoInicial * rendimento
        return (pesoCarcacaReal - carcacaInicial) / 15 * Double(animaisAbatidos)
    }

    var animaisPerdidos: Int { max(0, animaisIniciais - animaisAbatidos) }

    var taxaPerda: Double {
        animaisIniciais > 0 ? Double(animaisPerdidos) / Double(animaisIniciais) : 0
    }

    // MARK: - Custo

    var temCusto: Bool { custoReal > 0 }

    var custoPorArroba: Double? {
        guard custoReal > 0, arrobasProduzidasLote > 0 else { return nil }
        return custoReal / arrobasProduzidasLote
    }

    var custoPorKgGanho: Double? {
        guard custoReal > 0, ganhoTotalLote > 0 else { return nil }
        return custoReal / ganhoTotalLote
    }

    var custoPorAnimalDia: Double? {
        let base = diasReais * Double(animaisAbatidos)
        guard custoReal > 0, base > 0 else { return nil }
        return custoReal / base
    }

    var receita: Double? {
        guard precoArroba > 0, arrobasLote > 0 else { return nil }
        return precoArroba * arrobasLote
    }

    var margem: Double? {
        guard let receita, custoReal > 0 else { return nil }
        return receita - custoReal
    }

    /// Quanto do concentrado planejado foi de fato fornecido.
    var aderenciaConcentrado: Double? {
        guard concentradoPrevisto > 0, concentradoReal > 0 else { return nil }
        return concentradoReal / concentradoPrevisto
    }

    /// Desvio do custo em relação ao previsto.
    var desvioCusto: Double? {
        guard custoPrevisto > 0, custoReal > 0 else { return nil }
        return custoReal / custoPrevisto - 1
    }
}

File: web/src/nucleo/cicloEncerrado.ts
/**
 * Um lote que já foi abatido, com o que foi planejado e o que de fato aconteceu.
 *
 * Porte de `Core/Modelos/CicloEncerrado.swift`. É a memória da fazenda: cada
 * ciclo encerrado vira base de comparação e calibração para os lotes seguintes.
 */
import type { GrupoGenetico, SistemaCriacao } from "./classificacoes.js";

export interface CicloEncerrado {
  id: string;
  loteID: string;
  nome: string;
  grupoGenetico: GrupoGenetico;
  sistema: SistemaCriacao;

  // Plano, fotografado no momento do encerramento.
  dataInicio: Date;
  pesoInicial: number;
  ganhoMeta: number;
  pesoAlvo: number;
  animaisIniciais: number;
  pesoAcabamentoPlanejado: number;
  /** Consumo médio previsto, em kg de matéria seca por animal por dia. */
  consumoPrevistoDiario: number;
  /** Teores médios da dieta planejada (% da matéria seca). */
  ndtDietaMedia: number;
  pbDietaMedia: number;
  /** Concentrado previsto para o ciclo inteiro, em kg de matéria natural. */
  concentradoPrevisto: number;
  custoPrevisto: number;
  volumosoNome: string;
  energeticoNome: string;
  proteicoNome: string;

  // Realizado.
  dataAbate: Date;
  pesoFinalReal: number;
  animaisAbatidos: number;
  /** Peso médio de carcaça, em kg. Zero quando não informado. */
  pesoCarcacaReal: number;
  /** Concentrado realmente fornecido no ciclo, em kg de matéria natural. */
  concentradoReal: number;
  custoReal: number;
  /** Preço recebido por arroba. Zero quando não informado. */
  precoArroba: number;
  observacoes: string;
}

type Obrigatorios =
  | "loteID"
  | "nome"
  | "grupoGenetico"
  | "sistema"
  | "dataInicio"
  | "pesoInicial"
  | "ganhoMeta"
  | "pesoAlvo"
  | "animaisIniciais"
  | "pesoAcabamentoPlanejado"
  | "consumoPrevistoDiario"
  | "ndtDietaMedia"
  | "pbDietaMedia"
  | "concentradoPrevisto"
  | "custoPrevisto"
  | "volumosoNome"
  | "energeticoNome"
  | "proteicoNome"
  | "dataAbate"
  | "pesoFinalReal"
  | "animaisAbatidos";

export function criarCiclo(
  entrada: Pick<CicloEncerrado, Obrigatorios> & Partial<CicloEncerrado>,
): CicloEncerrado {
  return {
    id: crypto.randomUUID(),
    pesoCarcacaReal: 0,
    concentradoReal: 0,
    custoReal: 0,
    precoArroba: 0,
    observacoes: "",
    ...entrada,
  };
}

// ------------------------------------------------------- desempenho observado

/** Duração real do ciclo, em dias. */
export function diasReais(c: CicloEncerrado): number {
  return Math.max((c.dataAbate.getTime() - c.dataInicio.getTime()) / 86_400_000, 0);
}

export function diasPlanejados(c: CicloEncerrado): number {
  return c.ganhoMeta > 0 ? Math.max(0, c.pesoAlvo - c.pesoInicial) / c.ganhoMeta : 0;
}

export const ganhoTotalPorAnimal = (c: CicloEncerrado) =>
  Math.max(0, c.pesoFinalReal - c.pesoInicial);

export const ganhoTotalLote = (c: CicloEncerrado) =>
  ganhoTotalPorAnimal(c) * c.animaisAbatidos;

/** Ganho médio diário observado (kg/dia). */
export function ganhoRealDiario(c: CicloEncerrado): number {
  const dias = diasReais(c);
  return dias >= 1 ? ganhoTotalPorAnimal(c) / dias : 0;
}

/** Quanto do ganho planejado foi entregue. 1,0 = exatamente a meta. */
export function aderenciaGanho(c: CicloEncerrado): number {
  return c.ganhoMeta > 0 ? ganhoRealDiario(c) / c.ganhoMeta : 0;
}

/** Peso médio do animal ao longo do ciclo, usado nas contas de calibração. */
export const pesoMedioCiclo = (c: CicloEncerrado) => (c.pesoInicial + c.pesoFinalReal) / 2;

/** Rendimento de carcaça observado (fração). Nulo quando não informado. */
export function rendimentoReal(c: CicloEncerrado): number | null {
  if (!(c.pesoCarcacaReal > 0) || !(c.pesoFinalReal > 0)) return null;
  return c.pesoCarcacaReal / c.pesoFinalReal;
}

export const arrobasPorAnimal = (c: CicloEncerrado) =>
  c.pesoCarcacaReal > 0 ? c.pesoCarcacaReal / 15 : 0;

export const arrobasLote = (c: CicloEncerrado) => arrobasPorAnimal(c) * c.animaisAbatidos;

/** Arrobas produzidas no ciclo, descontando a carcaça de entrada. */
export function arrobasProduzidasLote(c: CicloEncerrado): number {
  const rendimento = rendimentoReal(c);
  if (rendimento === null) return 0;
  const carcacaInicial = c.pesoInicial * rendimento;
  return ((c.pesoCarcacaReal - carcacaInicial) / 15) * c.animaisAbatidos;
}

export const animaisPerdidos = (c: CicloEncerrado) =>
  Math.max(0, c.animaisIniciais - c.animaisAbatidos);

export function taxaPerda(c: CicloEncerrado): number {
  return c.animaisIniciais > 0 ? animaisPerdidos(c) / c.animaisIniciais : 0;
}

// ------------------------------------------------------------------- custo

export const temCusto = (c: CicloEncerrado) => c.custoReal > 0;

export function custoPorArroba(c: CicloEncerrado): number | null {
  const arrobas = arrobasProduzidasLote(c);
  if (!(c.custoReal > 0) || !(arrobas > 0)) return null;
  return c.custoReal / arrobas;
}

export function custoPorKgGanho(c: CicloEncerrado): number | null {
  const ganho = ganhoTotalLote(c);
  if (!(c.custoReal > 0) || !(ganho > 0)) return null;
  return c.custoReal / ganho;
}

export function custoPorAnimalDia(c: CicloEncerrado): number | null {
  const base = diasReais(c) * c.animaisAbatidos;
  if (!(c.custoReal > 0) || !(base > 0)) return null;
  return c.custoReal / base;
}

export function receita(c: CicloEncerrado): number | null {
  const arrobas = arrobasLote(c);
  if (!(c.precoArroba > 0) || !(arrobas > 0)) return null;
  return c.precoArroba * arrobas;
}

export function margem(c: CicloEncerrado): number | null {
  const r = receita(c);
  if (r === null || !(c.custoReal > 0)) return null;
  return r - c.custoReal;
}

/** Quanto do concentrado planejado foi de fato fornecido. */
export function aderenciaConcentrado(c: CicloEncerrado): number | null {
  if (!(c.concentradoPrevisto > 0) || !(c.concentradoReal > 0)) return null;
  return c.concentradoReal / c.concentradoPrevisto;
}

/** Desvio do custo em relação ao previsto. */
export function desvioCusto(c: CicloEncerrado): number | null {
  if (!(c.custoPrevisto > 0) || !(c.custoReal > 0)) return null;
  return c.custoReal / c.custoPrevisto - 1;
}
